Add a list of libraries that support __cuda_array_interface__ (#5104)
numba/numba: docs/source/cuda/cuda_array_interface.rst

- **shape**: ``(integer, ...)``

>   A tuple of `int` (or `long`) representing the size of each dimension.

Reviewer: ```suggestion
  A tuple of ``int`` (or ``long``) representing the size of each dimension.
```

> - **typestr**: `str`

Reviewer: ```suggestion
- **typestr**: ``str``
```

>   The type string.  This has the same definition as *typestr* in the

Reviewer: ```suggestion
  The type string.  This has the same definition as ``typestr`` in the
```

  `numpy array interface`_.

> - **data**: `(integer, boolean)`

Reviewer: ```suggestion
- **data**: ``(integer, boolean)``
```

  The **data** is a 2-tuple.  The first element is the data pointer

>   as a Python `int` (or `long`).  The data must be device-accessible.

Reviewer: ```suggestion
  as a Python ``int`` (or ``long``).  The data must be device-accessible.
```

>   For zero-size arrays, use `0` here.

Reviewer: ```suggestion
  For zero-size arrays, use ``0`` here.
```

>   The second element is the read-only flag as a Python `bool`.

Reviewer: ```suggestion
  The second element is the read-only flag as a Python ``bool``.
```

  Because the user of the interface may or may not be in the same context,
  the most common case is to use ``cuPointerGetAttribute`` with
  ``CU_POINTER_ATTRIBUTE_DEVICE_POINTER`` in the CUDA driver API (or the
  equivalent CUDA Runtime API) to retrieve a device pointer that
  is usable in the currently active context.

> - **version**: `integer`

Reviewer: ```suggestion
- **version**: ``integer``
```

  An integer for the version of the interface being exported.
  The current version is *2*.

The following are optional entries:

- **strides**: ``None`` or ``(integer, ...)``

  If **strides** is not given, or it is ``None``, the array is in

>   C-contiguous layout. Otherwise, a tuple of `int` (or `long`) is explicitly

Reviewer: ```suggestion
  C-contiguous layout. Otherwise, a tuple of ``int`` (or ``long``) is explicitly
```

  given for representing the number of bytes to skip to access the next
  element at each dimension.

- **descr**

  This is for describing more complicated types.  This follows the same
  specification as in the `numpy array interface`_.

- **mask**: ``None`` or object exposing the ``__cuda_array_interface__``

  If ``None`` then all values in **data** are valid. All elements of the mask
  array should be interpreted only as true or not true indicating which

>   elements of this array are valid. This has the same definition as *mask*

Reviewer: ```suggestion
  elements of this array are valid. This has the same definition as ``mask``
```

  in the `numpy array interface`_.

  .. note:: Numba does not currently support working with masked CUDA arrays

>             and will raise a `NotImplementedError` exception if one is passed

Reviewer: ```suggestion
            and will raise a ``NotImplementedError`` exception if one is passed
```

Versions 0 and 1 of the CUDA Array Interface neither clarified the
**strides** attribute for C-contiguous arrays nor specified the treatment for
zero-size arrays.

Interoperability
----------------

The following Python libraries have adopted the CUDA Array Interface:

- Numba
- `CuPy <https://docs-cupy.chainer.org/en/stable/reference/interoperability.html>`_
- `PyTorch <https://pytorch.org>`_
- `PyArrow <https://arrow.apache.org/docs/python/generated/pyarrow.cuda.Context.html#pyarrow.cuda.Context.buffer_from_object>`_
- `mpi4py <https://mpi4py.readthedocs.io/en/latest/overview.html#support-for-cuda-aware-mpi>`_
- `ArrayViews <https://github.com/xnd-project/arrayviews>`_

> - The **RAPIDS** stack:

Reviewer 2: Need this be bold font? it renders strangely.

Author: Ah...it's a leftover from previous style. Fixed.

> - The **RAPIDS** stack:

Author: ```suggestion
- The RAPIDS stack:
```

    - `cuDF <https://rapidsai.github.io/projects/cudf/en/0.11.0/10min-cudf-cupy.html>`_
    - `cuML <https://docs.rapids.ai/api/cuml/nightly/>`_
    - `cuSignal <https://github.com/rapidsai/cusignal>`_
    - `RMM <https://docs.rapids.ai/api/rmm/stable/>`_

> If your project is not on this list, please feel free to report to `Numba's issue tracker <https://github.com/numba/numba/issues>`_.

Author: ```suggestion
If your project is not on this list, please feel free to report it on the `Numba issue tracker <https://github.com/numba/numba/issues>`_.
```

Versions 0 and 1 of the CUDA Array Interface neither clarified the
**strides** attribute for C-contiguous arrays nor specified the treatment for
zero-size arrays.

> 

Reviewer 2: Whilst we're fixing up this doc, this entire section is incorrectly indented:
https://github.com/numba/numba/pull/5104/files#diff-017e6b3745f25f98ede11dfc0ede29caR26-R73

The paragraph after the list item needs indenting to the same level as the first character of the list item, i.e. 2 spaces.

Any chance you could fix this please whilst sorting out the issue below? Thanks!

Interoperability
----------------

The following Python libraries have adopted the CUDA Array Interface:

- **Numba**
- **CuPy** (`link <https://docs-cupy.chainer.org/en/stable/reference/interoperability.html>`_)
- **PyTorch**: (`link <https://pytorch.org>`_)
- **PyArrow**: (`link <https://arrow.apache.org/docs/python/generated/pyarrow.cuda.Context.html#pyarrow.cuda.Context.buffer_from_object>`_)
- **mpi4py**: (`link <https://mpi4py.readthedocs.io/en/latest/overview.html#support-for-cuda-aware-mpi>`_)
- **ArrayViews**: (`link <https://github.com/xnd-project/arrayviews>`_)
- The **RAPIDS** stack:
    - **cuDF**: (`link <https://rapidsai.github.io/projects/cudf/en/0.11.0/10min-cudf-cupy.html>`_)
    - **cuML**: (`link <https://docs.rapids.ai/api/cuml/nightly/>`_)
    - **cuSignal**: (`link <https://github.com/rapidsai/cusignal>`_)

>     - **RMM**: (`link <https://docs.rapids.ai/api/rmm/stable/>`_)

Reviewer 2: Could these not just be the links themselves opposed to adding `(link)` next to each?
e.g.:
```rst
- `CuPy <https://docs-cupy.chainer.org/en/stable/reference/interoperability.html>`_)
```

Author: Sure. I wonder if this also fixes @esc's comment on the sphinx errors?
